File: web/src/lib/telemetry.ts
export const STATUS_TOPIC = "showcase.status.v1";

const EVENT_TYPES = new Set([
  "session.connecting",
  "session.ready",
  "session.closed",
  "model.ready",
  "model.failed",
  "avatar.connecting",
  "avatar.ready",
  "avatar.failed",
  "knowledge.searching",
  "knowledge.matched",
  "knowledge.no_match",
  "tool.lookup.started",
  "tool.lookup.completed",
  "tool.request.started",
  "tool.request.completed",
  "tool.failed",
]);

const SOURCES = new Set(["runtime", "model", "avatar", "knowledge", "tool"]);
const STATUSES = new Set(["connecting", "ready", "closed", "running", "ok", "no_match", "failed"]);
const ALLOWED_KEYS = new Set([
  "schema_version",
  "sequence",
  "timestamp",
  "source",
  "type",
  "status",
  "label",
  "duration_ms",
]);

export type StatusEvent = {
  schema_version: 1;
  sequence: number;
  timestamp: string;
  source: "runtime" | "model" | "avatar" | "knowledge" | "tool";
  type: string;
  status: "connecting" | "ready" | "closed" | "running" | "ok" | "no_match" | "failed";
  label: string;
  duration_ms: number | null;
};

export function parseStatusEvent(payload: Uint8Array): StatusEvent | null {
  try {
    const parsed: unknown = JSON.parse(new TextDecoder().decode(payload));
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return null;
    const event = parsed as Record<string, unknown>;
    if (Object.keys(event).some((key) => !ALLOWED_KEYS.has(key))) return null;
    if (
      event.schema_version !== 1 ||
      typeof event.sequence !== "number" ||
      !Number.isSafeInteger(event.sequence) ||
      event.sequence < 1 ||
      typeof event.timestamp !== "string" ||
      Number.isNaN(Date.parse(event.timestamp)) ||
      typeof event.source !== "string" ||
      !SOURCES.has(event.source) ||
      typeof event.type !== "string" ||
      !EVENT_TYPES.has(event.type) ||
      typeof event.status !== "string" ||
      !STATUSES.has(event.status) ||
      typeof event.label !== "string" ||
      event.label.length < 1 ||
      event.label.length > 80 ||
      !(
        event.duration_ms === null ||
        (typeof event.duration_ms === "number" && event.duration_ms >= 0 && event.duration_ms <= 30000)
      )
    ) {
      return null;
    }
    return event as StatusEvent;
  } catch {
    return null;
  }
}

File: src/voice_agent/telemetry.py
from __future__ import annotations

import asyncio
import json
from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from typing import Any

TOPIC = "showcase.status.v1"
ALLOWED_EVENTS = {
    "session.connecting": ("runtime", "connecting", "Connecting session"),
    "session.ready": ("runtime", "ready", "Session ready"),
    "session.closed": ("runtime", "closed", "Session closed"),
    "model.ready": ("model", "ready", "Realtime model ready"),
    "model.failed": ("model", "failed", "Realtime model unavailable"),
    "avatar.connecting": ("avatar", "connecting", "Loading digital human"),
    "avatar.ready": ("avatar", "ready", "Digital human ready"),
    "avatar.failed": ("avatar", "failed", "Digital human unavailable"),
    "knowledge.searching": ("knowledge", "running", "Searching knowledge"),
    "knowledge.matched": ("knowledge", "ok", "Knowledge found"),
    "knowledge.no_match": ("knowledge", "no_match", "No verified answer"),
    "tool.lookup.started": ("tool", "running", "Checking reservation"),
    "tool.lookup.completed": ("tool", "ok", "Reservation lookup complete"),
    "tool.request.started": ("tool", "running", "Creating service request"),
    "tool.request.completed": ("tool", "ok", "Service request created"),
    "tool.failed": ("tool", "failed", "Business service unavailable"),
}


@dataclass(frozen=True, slots=True)
class StatusEvent:
    schema_version: int
    sequence: int
    timestamp: str
    source: str
    type: str
    status: str
    label: str
    duration_ms: int | None = None

    def to_json(self) -> str:
        return json.dumps(asdict(self), ensure_ascii=True, separators=(",", ":"))


class StatusPublisher:
    def __init__(self) -> None:
        self._participant: Any | None = None
        self._sequence = 0
        self._lock = asyncio.Lock()
        self.events: list[StatusEvent] = []

    async def activate(self, room: Any) -> None:
        async with self._lock:
            participant = room.local_participant
            for event in self.events:
                await participant.publish_data(event.to_json(), reliable=True, topic=TOPIC)
            self._participant = participant

    async def emit(self, event_type: str, *, duration_ms: int | None = None) -> StatusEvent:
        async with self._lock:
            if event_type not in ALLOWED_EVENTS:
                raise ValueError(f"Unsupported public event type: {event_type}")
            source, status, label = ALLOWED_EVENTS[event_type]
            self._sequence += 1
            event = StatusEvent(
                schema_version=1,
                sequence=self._sequence,
                timestamp=datetime.now(UTC).isoformat(timespec="milliseconds"),
                source=source,
                type=event_type,
                status=status,
                label=label,
                duration_ms=duration_ms,
            )
            self.events.append(event)
            if self._participant is not None:
                await self._participant.publish_data(event.to_json(), reliable=True, topic=TOPIC)
            return event
